### PR TITLE
Filters out private taxonomies from sidebar UI.

### LINKED ISSUE #2450: Gutenberg shows private taxonomies in Categories & Tags
It shouldn't. ;)

window

Prominent Words is a private taxonomy that our plugin creates and fills automatically through analysis of the content. The user being able to edit this is not intended behaviour.

### PR DESCRIPTION
**Pull request by BE-Webdesign**
Fixes #2450,

Could potentially break if the user does not have privileges to see the
edit context. Will need to test more.

**Testing Instructions**

1. Register a private taxonomy.
2. Load Gutenberg.
3. Verify that the taxonomy does not show up in the Categories & Tags portion of the sidebar.

### CODE REVIEW

File: lib/rest-api.php
Comment (aduth) [inline lib/rest-api.php]: Inaccurate description. Probably something about these being enhancements to existing endpoints? A bit unclear where the line should be drawn between `compat.php` and this one.

File: editor/sidebar/post-taxonomies/index.js
Comment (aduth) [inline editor/sidebar/post-taxonomies/index.js]: Maybe not here, but we should see about upgrading this component to using the `withAPIData` higher-order component.

https://github.com/WordPress/gutenberg/tree/master/components/higher-order/with-api-data

File: editor/sidebar/post-taxonomies/index.js
Comment (aduth) [inline editor/sidebar/post-taxonomies/index.js]: Minor: Lodash's `filter` handles this beautifully (coercing object to arrange, filtering on key):

```js
this.setState( {
	taxonomies: filter( taxonomies, { public: true } ),
} );
```

**Review by aduth — COMMENTED** [review]
Works well in my testing.

File: editor/sidebar/post-taxonomies/index.js
Comment (youknowriad) [reply editor/sidebar/post-taxonomies/index.js]: already done and merged, needs rebase :)

File: lib/rest-api.php
Comment (BE-Webdesign) [reply lib/rest-api.php]: Lol woops, copy pasta.

File: editor/sidebar/post-taxonomies/index.js
Comment (BE-Webdesign) [reply editor/sidebar/post-taxonomies/index.js]: Was thinking about using Lodash, this looks perfect!

File: editor/sidebar/post-taxonomies/index.js
Comment (BE-Webdesign) [reply editor/sidebar/post-taxonomies/index.js]: I will rebase and use the component.

File: lib/rest-api.php
Comment (BE-Webdesign) [reply lib/rest-api.php]: This should probably go into `compat.php`. Didn't realize what that was until now.

**Comment by BE-Webdesign** [comment]
@aduth, @youknowriad. This should be ready now. Rebased to use the new method Riad implemented, using higher order component. Does not fully work for authors, but that is tied to a larger problem.

**Comment by BE-Webdesign** [comment]
Fixes #2450,

Adds a public property to the taxonomy field, this is used to determine
whether the taxonomy is public or private. Private taxonomies will now
properly be hidden. Does not work 100% properly for Authors, this is a larger problem with
REST API.

**Comment by BE-Webdesign** [comment]
Anybody know why the test failed?

>Fatal error: Class 'WP_Block_Type' not found in /home/travis/build/WordPress/gutenberg/phpunit/class-wp-dummy-block-type.php on line 11

I don't think that has anything to do with this PR. Maybe re start the build?

**Comment by youknowriad** [comment]
@BE-Webdesign you should try to rebase, this probably means the build will fail on master if you merge, due to some other PR.

**Comment by BE-Webdesign** [comment]
Pretty sure I rebased against master. I will rebase again.

**Comment by BE-Webdesign** [comment]
That seemed to fix it, but I dunno why, it didn't bring in any new files etc.

**Comment by youknowriad** [comment]
computers 💻 are weird :)

**Comment by BE-Webdesign** [comment]
> computers 💻 are weird :)

Yup, this happens all too often 😅 .

**Review by youknowriad — APPROVED** [review]
Works great! I wonder if this fields should not be added by default to the Core Endpoint.

> Does not work 100% properly for Authors, this is a larger problem with REST API.

Could you detail the issue here and maybe create an issue/ticket for it?

Since we decided to start the API work here in Gutenberg and potentially move them to Core later, this looks good to me.

**Comment by BE-Webdesign** [comment]
> Could you detail the issue here and maybe create an issue/ticket for it?

#2545 is the ticket explaining most of it. Basically authors cannot view taxonomy data ever, due to permissions.

This PR, should magically work for authors, once #2545 is resolved, but they are somewhat separate issues.

**Comment by BE-Webdesign** [comment]
> I wonder if this fields should not be added by default to the Core Endpoint.

It is on the edit context, which is a restricted context, so this will only be visible to authenticated users. I am not sure what other conditional we would want to add to hide it. I think it should be fine as is. But I will wait for rmccue review, before merging.

**Comment by BE-Webdesign** [comment]
I think this problem is actually slightly more sinister than previously though. I have to verify but, I think private taxonomies will still display on a post page as long as the have a valid metabox callback set. So maybe we need to use that instead.

**Comment by pento** [comment]
I've created a [core ticket](https://core.trac.wordpress.org/ticket/42707) to expose the visibility settings for taxonomies.

Note: To mirror [Core behaviour](https://core.trac.wordpress.org/browser/trunk/src/wp-admin/edit-form-advanced.php?marks=260#L257), you need to test the `show_ui` setting, not the `public` setting.

**Comment by schlessera** [comment]
I did the following changes to update this PR:

- rebased to latest master
- added polyfill to backport https://core.trac.wordpress.org/ticket/42707
- adapted unit tests to check for visibility properties
- adapted post taxonomy component to hide on `visibility.show_ui: false`
- added test to ensure this works as expected

**Comment by danielbachhuber** [comment]
I rebased this against master and tested locally with the following:

```
add_action( 'init', function() {
    register_taxonomy( 'secrets', array( 'post' ), array(
        'label'             => 'Secrets',
        'public'            => true,
        'show_ui'           => false,
        'show_in_rest'      => true,
        'rest_base'         => 'secrets',
    ) );
});
```

Using this pull request, the "Secrets" tag UI isn't displayed as expected.

**Comment by mtias** [comment]
Nice to see this in!

**Pull request merged**
